labeling/service.go

```diff
@@ -0,0 +1,268 @@
+	"github.com/bluesky-social/indigo/pds"
+	"github.com/bluesky-social/indigo/repo"
+	"github.com/bluesky-social/indigo/repomgr"
+	util "github.com/bluesky-social/indigo/util"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	"github.com/labstack/echo/v4"
+	"github.com/labstack/echo/v4/middleware"
+	"github.com/whyrusleeping/go-did"
+	"gorm.io/gorm"
+)
+
+var log = logging.Logger("labelmaker")
+
+type Server struct {
+	db         *gorm.DB
+	cs         *carstore.CarStore
+	repoman    *repomgr.RepoManager
+	bgsSlurper *bgs.Slurper
+	levents    *events.LabelEventManager
+	echo       *echo.Echo
+	user       *LabelmakerRepoConfig
+	kwl        []KeywordLabeler
+}
+
+type LabelmakerRepoConfig struct {
+	handle     string
+	did        string
+	signingKey *did.PrivKey
+	userId     util.Uid
+}
+
+// In addition to configuring the service, will connect to upstream BGS and start processing events. Won't handle HTTP or WebSocket endpoints until RunAPI() is called.
+// 'useWss' is a flag to use SSL for outbound WebSocket connections
+func NewServer(db *gorm.DB, cs *carstore.CarStore, keyFile, repoDid, repoHandle, plcUrl string, useWss bool) (*Server, error) {
+
+	serkey, err := loadKey(keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	db.AutoMigrate(models.PDS{})
+
+	didr := &api.PLCServer{Host: plcUrl}
+	kmgr := indexer.NewKeyManager(didr, serkey)
+	levtman := events.NewLabelEventManager(events.NewMemLabelPersister())
+	repoman := repomgr.NewRepoManager(db, cs, kmgr)
+
+	user := &LabelmakerRepoConfig{
+		handle:     repoHandle,
+		did:        repoDid,
+		signingKey: serkey,
+		userId:     1,
+	}
+
+	var kl = KeywordLabeler{value: "meta", keywords: []string{"the", "bluesky", "atproto"}}
+
+	s := &Server{
+		db:      db,
+		repoman: repoman,
+		levents: levtman,
+		user:    user,
+		kwl:     []KeywordLabeler{kl},
+		// sluper configured below
+	}
+
+	// ensure that local labelmaker repo exists
+	// NOTE: doesn't need to have app.bsky profile and actor config, this is just expediant (reusing an existing helper function)
+	ctx := context.Background()
+	head, _ := s.repoman.GetRepoRoot(ctx, s.user.userId)
+	if head == cid.Undef {
+		log.Info("initializing labelmaker repo")
+		if err := s.repoman.InitNewActor(ctx, s.user.userId, s.user.handle, s.user.did, "Label Maker", pds.UserActorDeclCid, pds.UserActorDeclType); err != nil {
+			return nil, fmt.Errorf("creating labelmaker repo: %w", err)
+		}
+	} else {
+		log.Infof("found labelmaker repo: %s", head)
+	}
+
+	slurp := bgs.NewSlurper(db, s.handleBgsRepoEvent, useWss)
+	s.bgsSlurper = slurp
+
+	go levtman.Run()
+
+	return s, nil
+}
+
+func (s *Server) SubscribeBGS(ctx context.Context, bgsUrl string, useWss bool) {
+	// subscribe our RepoEvent slurper to the BGS, to receive incoming records for labeler
+	log.Infof("subscribing to BGS: %s (SSL=%v)", bgsUrl, useWss)
+	s.bgsSlurper.SubscribeToPds(ctx, bgsUrl, useWss)
```

Comment:
oh yeah probably could rename that method

Reply:
Yes, maybe `SubscribeToRepoHost`. But low priority.

Reply:
https://github.com/bluesky-social/indigo/issues/64 so this doesn't get lost